File: core/train.py
import sys
import json
import torch

sys.path.append("..")
from arch import architecture

# initialize arguments
eval_interval = int(sys.argv[1])
learning_rate = float(sys.argv[2])
vocab_size = 128

# initialize model and optimizer
model = architecture.GPTLanguageModel(vocab_size)
optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)
model.train()

i = 0
# one line is one batch in json format; keep reading while data is streaming in
for line in sys.stdin:
    # read from json and convert to tensor
    batch = json.loads(line)
    xb, yb = torch.tensor(batch["batch_x"]), torch.tensor(batch["batch_y"])

    # print iteration to logging pipeline 
    if i % eval_interval == 0:
         file_name = f"../logs/{i}.pth"
         torch.save(model.state_dict(), file_name)
         print(i, flush=True)
 
    # evaluate the loss and train
    logits, loss = model(xb, yb)
    optimizer.zero_grad(set_to_none=True)
    loss.backward()
    optimizer.step()
    
    #iterate
    i += 1

# save model
torch.save(model.state_dict(), f"../logs/{i}.pth")
print(i, flush=True)

File: core/train.js
#!/usr/bin/env node

import * as tf from '@tensorflow/tfjs-node';
import { createInterface } from 'readline';
import { GPTLanguageModel } from '../arch/nanogpt.js';
import { BigramLanguageModel } from '../arch/bigram.js';

// initialize arguments
const logInterval = parseInt(process.argv[2]);
const learningRate = parseFloat(process.argv[3]);
const modelDir = process.argv[4];
const vocabSize = 128;

// initialize model and optimizer
const model = new GPTLanguageModel(vocabSize); // can replace with BigramLanguageModel
const optimizer = tf.train.adam(learningRate);

async function train(line){
  // parse input
  const batch = JSON.parse(line);

  tf.tidy(() => {
    // TODO: create input and label tensors from batch
    // __start_solution__
    const x = tf.tensor2d(batch.xb, undefined, 'int32');
    const y = tf.tensor2d(batch.yb, undefined, 'int32');
    // __end_solution__
    
    // TODO: use the optimizer and the model loss function to train the model
    // __start_solution__
    optimizer.minimize(() => { return model.loss(x, y); });
    // __end_solution__
  });
};

async function main(){
  // create readline interface for stdin
  const rl = createInterface({ input: process.stdin });

  // main training loop
  let i = 0;
  for await (const line of rl) {
    // TODO: invoke the logger periodically (according to logInterval)
    // __start_solution__
    if (i % logInterval == 0) {
        await model.save(`../logs/${i}`);
        console.log(i);
    }
    // __end_solution
    
    // TODO: call the train function to train the model and iterate
    // __start_solution
    await train(line); 
    i++;
    // __end_solution
  }

  // save weights and log after training
  await model.save(`../logs/${i}`);
  console.log(i);

  // save to separate file for evaluations
  await model.save(`../${modelDir}`);
}

main();
